### PR TITLE
[WFLY-13435] / [WFLY-13509] / [WFLY-13434] / [WFLY-13319] DUP Refactoring for SecurityDomain resolution.

### PR DESCRIPTION
**Pull request by darranl**
This pull request depends upon https://github.com/wildfly/wildfly-core/pull/4276

As described in the WildFly Core PR this is predominantly about refactoring deployment unit processors so a security domain can be identified early and re-used by later deployment unit processors without repeating domain resolution logic.

The end result also means any virtual security domain created for MP JWT can be reused within an EJB deployment and a test case is added verifying the propagation of an identify established using JWT from the servlet container to the EJB container.

As described in the WildFly Core PR it will still be desirable to pull this resolution earlier so the security policy is defined before entering Phase.INSTALL but presently a deployment can be converted to a web application later in the INSTALL Phase.

https://issues.redhat.com/browse/WFLY-13435
https://issues.redhat.com/browse/WFLY-13509
https://issues.redhat.com/browse/WFLY-13434
https://issues.redhat.com/browse/WFLY-13319

### CODE REVIEW

**Comment by darranl** [comment]
Experimental jobs:
  https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/213201
  https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/213203

**Review by fjuma — APPROVED** [review]

**Comment by darranl** [comment]
The mixed domain test failures on https://ci.wildfly.org/buildConfiguration/WF_WildFlyCoreIntegrationExperiments/213521 can be ignored, this was a -Delytron run and we do not normally run mixed domain tests on that profile.

File: ejb3/src/main/java/org/jboss/as/ejb3/deployment/EJBSecurityDomainService.java [hunk 1]
```diff
@@ -36,39 +41,81 @@
...
-            final ClassLoader classLoader = module.getClassLoader();
-            registration = securityFunction.apply(deploymentName, classLoader);
+        final String deploymentName = deploymentUnit.getParent() == null ? deploymentUnit.getName()
+                : deploymentUnit.getParent().getName() + "." + deploymentUnit.getName();
+        final Module module = deploymentUnit.getAttachment(Attachments.MODULE);
+        final ClassLoader classLoader = module.getClassLoader();
+
+        ApplicationSecurityDomain applicationSecurityDomain = this.applicationSecurityDomain.getOptionalValue();
+        if (applicationSecurityDomain != null) {
+            BiFunction<String, ClassLoader, Registration> securityFunction = applicationSecurityDomain != null ? applicationSecurityDomain.getSecurityFunction() : null;
```
Comment (bstansberry) [inline ejb3/src/main/java/org/jboss/as/ejb3/deployment/EJBSecurityDomainService.java:79]: Nit: you don't need the null check any more as the line above ensures it's false.

Comment (darranl) [reply ejb3/src/main/java/org/jboss/as/ejb3/deployment/EJBSecurityDomainService.java:79]: True, I can clean that - there was an intermediate bug during development where that prior line didn't exist.

**Comment by darranl** [comment]
@chengfang / @fl4via / @rachmatowicz / @tadamski FYI this PR is handling DUP refactoring to better collaborate on the identification of the Elytron security domain to use and share the resolved security domain across deployments.

**Review by bstansberry — APPROVED** [review]

**Comment by bstansberry** [comment]
@chengfang @tadamski I'm going to go ahead and merge this one as it's blocking other critical work. If there are concerns we can follow up later.

**Pull request merged**